rust/arrow/src/array.rs

```diff
@@ -1659,13 +1710,80 @@ mod tests {
         field_types.push(Field::new("a", DataType::Boolean, false));
         field_types.push(Field::new("b", DataType::Int64, false));
         let struct_array_data = ArrayData::builder(DataType::Struct(field_types))
+            .len(4)
+            .add_child_data(boolean_data.clone())
+            .add_child_data(int_data.clone())
+            .build();
+        let struct_array = StructArray::from(struct_array_data);
+
+        assert_eq!(boolean_data, struct_array.column(0).data());
+        assert_eq!(int_data, struct_array.column(1).data());
+    }
+
+    #[test]
+    fn test_struct_array_slice() {
+        let mut null_bits: [u8; 1] = [0; 1];
+        bit_util::set_bit(&mut null_bits, 0);
+        bit_util::set_bit(&mut null_bits, 2);
+        bit_util::set_bit(&mut null_bits, 4);
+        let null_bit_buffer = Buffer::from(null_bits);
+
+        let boolean_data = ArrayData::builder(DataType::Boolean)
+            .len(5)
+            .add_buffer(Buffer::from(
+                [false, false, false, true, false].to_byte_slice(),
+            ))
+            .null_bit_buffer(null_bit_buffer.clone())
+            .build();
+        let int_data = ArrayData::builder(DataType::Int32)
+            .len(5)
+            .add_buffer(Buffer::from([42, 28, 19, 31, 54].to_byte_slice()))
+            .null_bit_buffer(null_bit_buffer.clone())
+            .build();
+
+        let mut field_types = vec![];
+        field_types.push(Field::new("a", DataType::Boolean, false));
+        field_types.push(Field::new("b", DataType::Int32, false));
+        let struct_array_data = ArrayData::builder(DataType::Struct(field_types))
+            .len(5)
             .add_child_data(boolean_data.clone())
             .add_child_data(int_data.clone())
+            .null_bit_buffer(null_bit_buffer)
             .build();
         let struct_array = StructArray::from(struct_array_data);
 
+        assert_eq!(5, struct_array.len());
         assert_eq!(boolean_data, struct_array.column(0).data());
         assert_eq!(int_data, struct_array.column(1).data());
+
+        let sliced_array = struct_array.slice(2, 3);
+        let sliced_array = sliced_array.as_any().downcast_ref::<StructArray>().unwrap();
+        assert_eq!(3, sliced_array.len());
+        assert_eq!(2, sliced_array.offset());
+        assert_eq!(1, sliced_array.null_count());
+        assert!(sliced_array.is_valid(0));
+        assert!(sliced_array.is_null(1));
+        assert!(sliced_array.is_valid(2));
+
+        let sliced_c0 = sliced_array.column(0);
+        let sliced_c0 = sliced_c0.as_any().downcast_ref::<BooleanArray>().unwrap();
+        assert_eq!(3, sliced_c0.len());
+        assert_eq!(2, sliced_c0.offset());
+        assert!(sliced_c0.is_valid(0));
```

Comment:
Do you mind instead building the arrays with the `From<Vec<>>` builder, so they get their own butmaps? I think sharing the bitmap of the struct array shadows the intention that we should test the validity of the struct's column by its own bitmap instead of the struct's one.

Reply:
We can pass different bitmaps for different columns even with the current approach. Let me update the PR.